url86e9087e81_JohnPease_557_HW3_tgz-pJ8-uwm_cs557_Sample4J8/src/uwm_cs557/MiniDB.java

```diff
@@ -0,0 +1,229 @@
+   * @return The dbBlockID of the new block is returned or -1 if there no room for an addtional block.
+   */ 
+  public int addBlockToDB( ) throws MiniDB_Exception{
+    int newBlockID = (int)-1;
+
+    for (int i=0; i<MiniDB_Constants.DB_MAX_BLOCKS && newBlockID==-1; i++){
+      if( !dbBlockUsed[i] ){
+	newBlockID = (int)i;
+	dbBlockUsed[i] = true;
+      }
+    }
+
+    if( newBlockID == -1 ) throw new MiniDB_Exception( MiniDB_Exception.DB_FULL );
+
+    //MiniDB.log( "BufMgr.addBlockToDB, new block ID = " + newBlockID );
+
+    return newBlockID;
+  }
+
+
+  public int numBlocksInDB(){
+    int cnt = 0;
+    for (int i=0; i<MiniDB_Constants.DB_MAX_BLOCKS; i++) if( dbBlockUsed[i] ) cnt++;
+
+    return cnt;
+  }
+
+
+  protected static String indexNameKey(  String indexedFilename, String fieldName ){
+    return indexedFilename + "#" + fieldName;
+  }
+
+  public BPlusTree loadExistingIndex( String indexedFilename,
+				      String indexedFieldname ) throws MiniDB_Exception{
+    BPlusTree bpt = null;
+    BPlusTreeHeadBlock indexHeadBlock = null;
+
+    int headBlockID = catalog.getIndexHeadBlockID( indexNameKey(indexedFilename, indexedFieldname) );
+    MiniDB_Block blk = bufMgr.pin_block( headBlockID );
+
+    if( blk != null ){
+      bpt = new BPlusTree(this);
+      indexHeadBlock = new BPlusTreeHeadBlock( blk );
+      bpt.init_existing( indexedFilename, indexedFieldname, indexHeadBlock );
+    }
+    else{
+      // something went wrong. no room in buffer mgr?
+    }
+    
+    return bpt;
+  }
+
+  public BPlusTree createNewBPlusTreeIndex( String indexedFileName,
+					    String indexedFieldName, 
+					    int indexedFieldSize) throws MiniDB_Exception{
+    BPlusTreeHeadBlock indexHeadBlock = null;
+    BPlusTree bpt = null;
+    String indexName = indexNameKey(indexedFileName, indexedFieldName);
+    int headBlockID = catalog.getIndexHeadBlockID( indexName );
+
+    if( headBlockID != -1 ){
+      // There is already an index on this field in the DB
+    }
+    else{
+      // add a block to the DB for the head block
+      headBlockID = this.addBlockToDB(); 
+
+      if( headBlockID != -1 ){
+	MiniDB_Block block = bufMgr.pin_block( headBlockID );
+	
+	if( block == null ){
+	  // no room in the buffer pool. Should delete the head block
+	}
+	else{
+	  indexHeadBlock = new BPlusTreeHeadBlock( block );
+	  indexHeadBlock.setBlockType( MiniDB_Block.BLOCK_TYPE_BPT_HEADER );
+	  catalog.addIndex( indexName, headBlockID );
+	  
+	  bpt = new BPlusTree(this);
+	  bpt.init_new(indexedFileName, 
+		       indexedFieldName, 
+		       indexedFieldSize, 
+		       indexHeadBlock );
+
+	}
+      }
+      else{
+	// no room in the database
+      }
+    }
+
+    return bpt;
+  }
+
+  public void writeDBToOSFile(){
+
+    bufMgr.flushDirty();
+
+    try {
+      FileOutputStream fout = new FileOutputStream(osFilename());
+      ObjectOutputStream oos = new ObjectOutputStream(fout);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *oos*. The resource is referenced at lines 183, 184, 185. The resource is closed at line 186. There are other execution paths that do not contain closure statements, for example, when ObjectOutputStream.writeObject throws an exception. Ensure *oos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).